### ISSUE TITLE
False positive with nested functions: warning CS0162: Unreachable code detected

### ISSUE THREAD

**Issue by AArnott** [dotnet/roslyn#51010]
**Version Used**: VS 16.10 Preview 1 (31004.24.main)

**Steps to Reproduce**:

[Sharplab](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYZQAs6AsgBQCU+hxTAEgJZ0DcjTR3PUAdi5IexPkwDGAewB2AZwAumOAAZ0AWzkBzdAF50AImYBTADampB4aN4ib5dGzp6AfMoCc1TVtrXiAX2R/IA=)

```cs
using System;
public class C {
    public void M() {
        Hi();
        
        return;
        
        const string msg = "Hello";
        
        void Hi() => Console.WriteLine(msg);
    }
}
```

**Expected Behavior**:

No warning.

**Actual Behavior**:

This warning appears on the line defining the constant, which is indeed useful code.
```
warning CS0162: Unreachable code detected
```

Also, if I put the `return;` statement *below* the `const` declaration, the C# lang service fades the declaration as if it is an unused const, even though it is.
IDE0059 then shows up and offers to remove a redundant assignment, but its code fix then introduces errors and more warnings.

**Comment by manfred-brands** [comment]
Isn't this a duplicate of #41429 ?

**Comment by shuebner** [comment]
@manfred-brands it seems so, yes. But one of the issues should remain open until this is actually fixed.
This very issue could be re-opened as well because the FP exists at least with SDK 6.0.405
